zerver/lib/mention.py

```diff
@@ -269,15 +270,30 @@ def message_has_topic_wildcards(self) -> bool:
 
     def init_user_group_data(self, realm_id: int, content: str) -> None:
         self.user_group_name_info: dict[str, NamedUserGroup] = {}
-        self.user_group_members: dict[int, set[int]] = {}
+        self.user_group_members: dict[int, set[int]] = defaultdict(set)
         user_group_names = possible_user_group_mentions(content)
         if user_group_names:
-            for group in NamedUserGroup.objects.filter(
+            named_user_groups = NamedUserGroup.objects.filter(
                 realm_id=realm_id, name__in=user_group_names
-            ):
-                self.user_group_name_info[group.name.lower()] = group
-                self.user_group_members[group.id] = set(
-                    get_recursive_group_members(group).values_list("id", flat=True)
+            )
+            # When there are no matching user_group_names (e.g non-existent user_group_names),
+            # then no need to execute the relatively expensive recursive CTE query
+            # triggered by get_root_id_annotated_recursive_subgroups_for_groups
+            if len(named_user_groups) == 0:
```

Comment:
I deliberatly chose to use `len()` over `exists()`, while the later is typically faster, it executes extra query to check the length.